File: docs/Sword2offer/CH2-编程题目-基础部分/ES6/jumpSteps.js
let jump = function () {
  let steps = [1, 2],
    cached = {}
  function updateTheory (step, lastTheory) {
    return lastTheory.map((item) => {
      // 不使用.push(step)。因为push这个api没有返回数组。
      return [step].concat(item)
    })
  }
  function findTheory (target) {
    let theory = []
    if (target <= 0) {
      cached[target] = [[]]
      return cached[target]
    }
    if (cached[target]) {
      return cached[target]
    }
    for (let i = 0; i < steps.length; i++) {
      let step = steps[i]
      let newTarget = target - step
      let newTheory = []
      if (newTarget >= 0) {
        // step需要和上一步策略中每一种可能性相互连接
        newTheory = updateTheory(step, findTheory(newTarget))
        // 连接step=1和step=2两种策略结果
        theory = theory.concat(newTheory)
      }
    }
    cached[target] = theory
    return theory
  }
  return function (target) {
    if (cached[target]) {
      return cached[target]
    } else {
      let result = findTheory(target)
      return result
    }
  }
}

let jumpSteps = jump()
// 边界测试
console.log(jumpSteps(0))
console.log(jumpSteps(1))
console.log(jumpSteps(2))
console.log(jumpSteps(3))
// 功能测试
console.log('5', jumpSteps(5))
console.log('10', jumpSteps(10))

File: docs/Sword2offer/CH2-编程题目-基础部分/Python/jumpSteps.py
class Jump:
  def __init__(self, steps):
    self.cached = {}
    self.steps = steps

  def __update_theory(self, step, last_theory):
    current_theory = []
    for theory in last_theory:
      theory.append(step)
      current_theory.append(theory)
    return current_theory
  
  def __find_theory(self, target):
    theory = []
    if target <= 0:
      self.cached[target] = [[]]
      return [[]]
    for step in self.steps:
      new_theory = []
      new_target = target - step
      if new_target >= 0:
        new_theory = self.__update_theory(step, self.__find_theory(new_target))
        theory += new_theory
    self.cached[target] = theory
    return theory
  
  def find_theory(self, target):
    if target in self.cached:
      return self.cached[target]
    else:
      return self.__find_theory(target)


jump = Jump([1, 2])
print(jump.find_theory(0))
print(jump.find_theory(1))
print(jump.find_theory(2))
# print('性能测试')
print(jump.find_theory(3))
print(jump.find_theory(5))
print(jump.find_theory(10))
print(jump.find_theory(100))